janus_collector/src/lib.rs

```diff
@@ -0,0 +1,866 @@
+    BadContentType(Option<HeaderValue>),
+    #[error("Codec error: {0}")]
+    Codec(#[from] prio::codec::CodecError),
+    #[error("Aggregate share decoding error")]
+    AggregateShareDecode,
+    #[error("Expected two aggregate shares, got {0}")]
+    AggregateShareCount(usize),
+    #[error("VDAF error: {0}")]
+    Vdaf(#[from] prio::vdaf::VdafError),
+    #[error("HPKE error: {0}")]
+    Hpke(#[from] janus_core::hpke::Error),
+    #[error("Timed out waiting for collection to finish")]
+    CollectPollTimeout,
+}
+
+static COLLECTOR_USER_AGENT: &str = concat!(
+    env!("CARGO_PKG_NAME"),
+    "/",
+    env!("CARGO_PKG_VERSION"),
+    "/",
+    "collector"
+);
+
+/// Authentication configuration for communication with the leader aggregator.
+#[derive(Derivative)]
+#[derivative(Debug)]
+#[non_exhaustive]
+pub enum Authentication {
+    /// Beearer token authentication, via the `DAP-Auth-Token` header.
+    BearerToken(#[derivative(Debug = "ignore")] String),
+}
+
+/// The PPM collector's view of task parameters.
+#[derive(Derivative)]
+#[derivative(Debug)]
+pub struct CollectorParameters {
+    /// Unique identifier for the task.
+    task_id: TaskId,
+    /// The base URL of the leader's aggregator API endpoints.
+    #[derivative(Debug(format_with = "std::fmt::Display::fmt"))]
+    leader_endpoint: Url,
+    /// The authentication information needed to communicate with the leader aggregator.
+    authentication: Authentication,
+    /// HPKE configuration and public key used for encryption of aggregate shares.
+    hpke_config: HpkeConfig,
+    /// HPKE private key used to decrypt aggregate shares.
+    #[derivative(Debug = "ignore")]
+    hpke_private_key: HpkePrivateKey,
+    /// Parameters to use when retrying HTTP requests.
+    http_request_retry_parameters: ExponentialBackoff,
+    /// Parameters to use when waiting for a collect job to be processed.
+    collect_poll_wait_parameters: ExponentialBackoff,
+}
+
+impl CollectorParameters {
+    /// Creates a new set of collector task parameters.
+    pub fn new(
+        task_id: TaskId,
+        leader_endpoint: Url,
+        authentication_token: String,
+        hpke_config: HpkeConfig,
+        hpke_private_key: HpkePrivateKey,
+    ) -> CollectorParameters {
+        CollectorParameters::new_with_backoff(
+            task_id,
+            leader_endpoint,
+            authentication_token,
+            hpke_config,
+            hpke_private_key,
+            http_request_exponential_backoff(),
+            ExponentialBackoff {
+                initial_interval: StdDuration::from_secs(15),
+                max_interval: StdDuration::from_secs(300),
+                multiplier: 1.5,
+                max_elapsed_time: None,
+                ..Default::default()
+            },
+        )
+    }
+
+    /// Creates a new set of collector task parameters with non-default HTTP request retry
+    /// parameters.
+    #[allow(clippy::too_many_arguments)]
```

Comment:
optional: instead of having two "constructor" functions, we could use a builder-like pattern. I suppose `new` would stay as it is (since every parameter to `new` is a required parameter w/ no useful default value); but then we could introduce `.with_http_request_backoff(self, ExponentialBackoff)` & `.with_collect_poll_backoff(self, ExponentialBackoff)` which consume an existing `CollectorParameters` and update the given field.

The upside is that we don't have to create `2^n` constructor variants for every set of fields we decide to independently support -- we have a single constructor variant & `O(n)` "builder" methods.

Reply:
+1 to this, especially if we can find some nice macros that generate the builder functions based on a struct declaration.